text/0000-introduce empty ranges.md

```diff
@@ -0,0 +1,431 @@
+
+The current implementation also provides an `.is_infinite()` function to test whether the iterator will indefinitely return values. This function could be currently be in use in to *avoid* iterating over Ranges that might be infinite or to produce appropriate errors. Since this RFC reduces the number of cases in which Ranges are infinite, this might produce side effects in existing code even if that code does not purposely use infinite Ranges as control structures like the first example. This could for example prevent certain error messages from getting displayed that were used in combination with such tests. It is on the other hand fair to state that a search of public Pony code on Github did not yield any uses of `.is_infinite()` outside test code associated with tests for the Range implementation. This could be taken as a sign that infinite Ranges and guarding code against their effects are not widely used. It is also worth noting that `.is_infinite()` will still return true for all remaining infinite Range cases so that any use as a guard should likely remain intact.
+
+# Alternatives
+
+An alternative way to avoid infinitely iterating Ranges would be a reinterpretation of the user-provided parameters in the currently *infinite* cases. For example, the currently infinite `Range(0, 10, -1)` could be re-arranged to mean `Range(10, 0, -1)`. An argument against transforming Range(0, 10, -1) into Range(10, 0, -1) is that Range evaluates expressions using the provided parameters in their meaning of `min` and `max` or `start` and `end`. In conjunction with the step parameter there is a directionality (trajectory) which does not only control the numbers of the sequence but also the order within the sequence. Instead of `0, 1, .., 9` the Range would produce  `10, 9, .., 1` -- different numbers in a different order.
+Furthermore, having empty Ranges when using `Range` with calculated, non-literal bounds also has an *expressive* use (see CSV reader example in Motivation section) which the reinterpretation of the parameters to always return a non-empty Range would preclude.
+
+Another fair alternative is to leave everything as it is. Pony has been in fairly wide-spread use academically and to some degree even commercially, without anyone reporting the issues that for the author were the reason to start working on this RFC -- which was hanging (eventually running out of memory) Pony code. One could therefore speculate that the side effects and instances in which currently infinite Ranges are created are sufficiently documented and known by the majority of users of Pony.
+
+# Unresolved questions
```

Comment:
Also, perhaps this RFC should also apply these changes to the [`Reverse`](https://stdlib.ponylang.io/collections-Reverse/) class that has the same problems as `Range`, and updating its documentation too.

@stefandd, the suggested changes are very good. The fact that pony has a `Reverse` iterator may question why we allow negative `inc`, but at least your proposal will prevent some type of bugs.

Reply:
My initial thought was "surely this would affect `Reverse` implicitly" as I thought `Reverse` was implemented via `Range` but that is not true. I am in favor of having `Range` and `Reverse` mirror one another since even the docstring for `Reverse` states itself as "(i.e. the reverse of `Range`)." Whether that should be rolled into this RFC or not, I have no strong opinion on. The one caveat for why `Reverse` is not a direct call to `Range` underneath is that the `Number` parameter can be unsigned. There would have to have a `direction` argument to `Range` which switches `+ increment` or `- increment`.